File: cpp/45_jump/jump.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int jump(vector<int>& nums) {
    int maxPos = 0, n = nums.size(), end = 0, step = 0;
    for (int i = 0; i < n - 1; ++i) {
      if (maxPos >= i) {
        maxPos = max(maxPos, i + nums[i]);
        if (i == end) {
          end = maxPos;
          ++step;
        }
      }
    }
    return step;
  }
};

int main() {
  vector<tuple<vector<int>, int>> tests{
      {{2, 3, 1, 1, 4}, 2},
      {{2, 3, 0, 1, 4}, 2},
  };

  for (auto& [nums, ans] : tests) {
    assert(Solution().jump(nums) == ans);
  }
}

File: rust/45_jump/jump.rs
struct Solution;

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let mut max_pos = 0;
        let mut end = 0;
        let mut steps = 0;
        for i in 0..nums.len() - 1 {
            if max_pos >= i {
                max_pos = max_pos.max(i + nums[i] as usize);
                if i == end {
                    end = max_pos;
                    steps += 1;
                }
            }
        }
        steps
    }
}

fn main() {
    let tests = vec![(vec![2, 3, 1, 1, 4], 2), (vec![2, 3, 0, 1, 4], 2)];

    for (nums, ans) in tests {
        assert_eq!(Solution::jump(nums), ans);
    }
}
